sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -0,0 +1,133 @@
+          }
+        }
+      } else {
+        collectionSize.incRefCount()
+      }
+    }
+  }
+}
+
+case class GpuElementAt(left: Expression, right: Expression)
+  extends GpuBinaryExpression with ExpectsInputTypes {
+
+  override lazy val dataType: DataType = left.dataType match {
+    case ArrayType(elementType, _) => elementType
+    case MapType(_, valueType, _) => valueType
+  }
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    (left.dataType, right.dataType) match {
+      case (arr: ArrayType, e2: IntegralType) if (e2 != LongType) =>
+        Seq(arr, IntegerType)
+      case (MapType(keyType, valueType, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(keyType, e2) match {
+          case Some(dt) => Seq(MapType(dt, valueType, hasNull), dt)
+          case _ => Seq.empty
+        }
+      case (l, r) => Seq.empty
+    }
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (_: ArrayType, e2) if e2 != IntegerType =>
+        TypeCheckResult.TypeCheckFailure(s"Input to function $prettyName should have " +
+          s"been ${ArrayType.simpleString} followed by a ${IntegerType.simpleString}, but it's " +
+          s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
+      case (MapType(e1, _, _), e2) if (!e2.sameType(e1)) =>
+        TypeCheckResult.TypeCheckFailure(s"Input to function $prettyName should have " +
+          s"been ${MapType.simpleString} followed by a value of same key type, but it's " +
+          s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
+      case (e1, _) if (!e1.isInstanceOf[MapType] && !e1.isInstanceOf[ArrayType]) =>
+        TypeCheckResult.TypeCheckFailure(s"The first argument to function $prettyName should " +
+          s"have been ${ArrayType.simpleString} or ${MapType.simpleString} type, but its " +
+          s"${left.dataType.catalogString} type.")
+      case _ => TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  // Eventually we need something more full featured like
+  // GetArrayItemUtil.computeNullabilityFromArray
+  override def nullable: Boolean = true
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+  override def doColumnar(lhs: Scalar, rhs: GpuColumnVector): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: Scalar): ColumnVector = {
+    lhs.dataType match {
+      case _: ArrayType => {
+        if (rhs.isValid) {
+          if (rhs.getInt > 0) {
+            // SQL 1-based index
+            lhs.getBase.extractListElement(rhs.getInt - 1)
+          } else if (rhs.getInt == 0) {
+            throw new ArrayIndexOutOfBoundsException("SQL array indices start at 1")
+          } else {
+            lhs.getBase.extractListElement(rhs.getInt)
```

Comment:
Does this have similar ANSI issues as https://github.com/NVIDIA/spark-rapids/pull/2350?

Should we refactor the code so that the core part of GpuGetArrayItem and GpuGetMapValue are in static methods so we can define GetElementAt in terms of them?

Reply:
Yes, GetArrayItem and GetMapValue both have ANSI mode problem.
For GetMapValue, I think we can extract the common code for core part.
For GetArrayItem, there's a little difference: element_at requires 1 as the start index, while GetArrayItem uses 0. There might be a parameter like "start_zero: Boolean" for its core function.

I'll take a more close look at it.

Reply:
Hi @revans2 , I refactored the part and move common codes into static methods, could you help take a look ? Thanks!